### PR TITLE
L3DAS22 enhancement  Recipe

### PR DESCRIPTION
**Pull request by neillu23**

### CODE REVIEW

File: egs2/l3das22/enh1/local/l3das22_multi_channel.sh [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+fi
+
+# Set bash to 'debug' mode, it will exit on :
+# -e 'error', -u 'undefined variable', -o ... 'error in pipeline', -x 'print commands',
+set -e
+set -u
+set -o pipefail
+
+# configs
+nj=16
```
Comment (Emrys365) [inline egs2/l3das22/enh1/local/l3das22_multi_channel.sh:19]: If this is a parsable argument, it is better to add a `help_message` string like in [egs2/chime4/enh1/local/data.sh#L13-L30](https://github.com/espnet/espnet/blob/master/egs2/chime4/enh1/local/data.sh#L13-L30).

**Review by Emrys365 — COMMENTED** [review]
Thanks! I just made some minor comments.

**Comment by Emrys365** [comment]
Could you also add your recipe to [egs2/README.md](https://github.com/espnet/espnet/blob/master/egs2/README.md)?

File: egs2/l3das22/enh1/local/normalize_transcript.pl [hunk 2]
```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/env perl
```
Comment (sw005320) [inline egs2/l3das22/enh1/local/normalize_transcript.pl:1]: is it needed?

**Review by sw005320 — COMMENTED** [review]
This recipe should have a special evaluation (scoring) script prepared by an organizer (or by making the exact same scoring scheme by ourselves).

**Comment by neillu23** [comment]
Thanks for the review and comments! I have addressed the comments and made two major changes to the PR:
1. Add the testing data from the L3DAS challenge.
2. Add metric.sh for calculating the challenge score.

File: egs2/l3das22/enh1/local/metric.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+import numpy as np
+import soundfile as sf
+import torchaudio
+import torch
+from tqdm import tqdm
+
+import jiwer
+import numpy as np
+from pystoi import stoi
+import transformers
```
Comment (sw005320) [inline egs2/l3das22/enh1/local/metric.py:22]: Is it OK?
I think we need to install something.

File: egs2/l3das22/enh1/local/metric.sh [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  echo "Usage: local/metric.sh <predicted scp> <target scp>"
+  exit 1;
+fi
+
+if [ $# -gt 1 ]; then
+	predicted_path=$1
+	target_path=$2
+  python local/metrics.py --predicted_path ${predicted_path} --target_path ${target_path}
+else
+	python local/metrics.py
```
Comment (sw005320) [inline egs2/l3das22/enh1/local/metric.sh:16]: The indent looks wrong.

File: egs2/l3das22/enh1/local/metric.py (hunk 3 above)
Comment (neillu23) [reply egs2/l3das22/enh1/local/metric.py:22]: Yes, we need to install jiwer (WER), pystoi (STOI), and transformers (ASR).
I can add an installation part to the metric.sh.

File: egs2/l3das22/enh1/local/metric.sh (hunk 4 above)
Comment (neillu23) [reply egs2/l3das22/enh1/local/metric.sh:16]: Thanks for pointing it out. This didn't happen in my code viewer.

File: egs2/l3das22/enh1/local/metric.sh [hunk 5]
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env bash
+
+[ -f ./path.sh ] && . ./path.sh
+. parse_options.sh || exit 1;
+
+pip show -f jiwer >/dev/null || pip install jiwer
+pip show -f pystoi >/dev/null || pip install pystoi
```
Comment (Emrys365) [inline egs2/l3das22/enh1/local/metric.sh:7]: I think `pystoi` is ensured to be installed in setup.py

Comment (neillu23) [reply egs2/l3das22/enh1/local/metric.sh:7]: Thanks! Let me remove it.

File: egs2/l3das22/enh1/local/data.sh [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    echo -e "$(date '+%Y-%m-%dT%H:%M:%S') (${fname}:${BASH_LINENO[0]}:${FUNCNAME[1]}) $*"
+}
+
+
+help_message=$(cat << EOF
+Usage: $0 
+  optional argument:
+    None
+EOF
+)
```
Comment (ftshijt) [inline egs2/l3das22/enh1/local/data.sh:18]: Maybe just a warning message for calling with other arguments? It seems not a large problem

File: egs2/l3das22/enh1/local/data.sh [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+if [ $# -ne 0 ]; then
+    echo "${help_message}"
+    exit 1;
+fi
+
+if [ ! -e "${L3DAS22}" ] ; then
+    log "
+    Please fill the value of 'L3DAS22' in db.sh
+    The 'L3DAS22' (https://www.kaggle.com/l3dasteam/l3das22) directory 
+    should at least contain the task1 dataset:
```
Comment (ftshijt) [inline egs2/l3das22/enh1/local/data.sh:39]: In that case, does it possible to directly download the data from kaggle API?

File: egs2/l3das22/enh1/local/l3das22_data_prep.sh [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+# Set bash to 'debug' mode, it will exit on :
+# -e 'error', -u 'undefined variable', -o ... 'error in pipeline', -x 'print commands',
+set -e
+set -u
+set -o pipefail
+
+L3DAS22=$1
+
+tmpdir=data/temp
```
Comment (ftshijt) [inline egs2/l3das22/enh1/local/l3das22_data_prep.sh:29]: `mktemp` would be better in the case, and please also use trap to delete the corresponding folder after the preparation.

File: egs2/l3das22/enh1/local/l3das22_multi_channel.sh [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+dev_dir=${L3DAS22}/L3DAS22_Task1_dev
+test_dir=${L3DAS22}/L3DAS22_Task1_test
+train_dir100=${L3DAS22}/L3DAS22_Task1_train100
+train_dir360_1=${L3DAS22}/L3DAS22_Task1_train360_1
+train_dir360_2=${L3DAS22}/L3DAS22_Task1_train360_2
+
+
+echo "Building multi-channel data"
+
+tmpdir=data/temp
```
Comment (ftshijt) [inline egs2/l3das22/enh1/local/l3das22_multi_channel.sh:58]: same for here, `mktemp` seems to fit the context better

File: egs2/l3das22/enh1/local/metric.sh [hunk 10]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env bash
+
+[ -f ./path.sh ] && . ./path.sh
+. parse_options.sh || exit 1;
+
+pip show -f jiwer >/dev/null || pip install jiwer
+pip show -f transformers >/dev/null || pip install transformers
```
Comment (ftshijt) [inline egs2/l3das22/enh1/local/metric.sh:7]: We may add a note that these evaluation metrics are from the official requirements?

File: egs2/l3das22/enh1/pyscripts [hunk 11]
```diff
@@ -0,0 +1 @@
+../../TEMPLATE/asr1/pyscripts
```
Comment (ftshijt) [inline egs2/l3das22/enh1/pyscripts:1]: though may not be different, by use `enh1/pyscripts` would be better (just in case future differences)

**Review by ftshijt — COMMENTED** [review]
Looks very cool! some minor suggestion

File: egs2/l3das22/enh1/local/l3das22_data_prep.sh (hunk 8 above)
Comment (neillu23) [reply egs2/l3das22/enh1/local/l3das22_data_prep.sh:29]: Since the files under the temp directory are used in the following recipe (local/l3das22_multi_channel.sh), do you think we could keep using this directory for passing the files? (or we could move mktemp to local/data.sh)

File: egs2/l3das22/enh1/local/data.sh (hunk 6 above)
Comment (neillu23) [reply egs2/l3das22/enh1/local/data.sh]: Added.

File: egs2/l3das22/enh1/local/data.sh (hunk 7 above)
Comment (neillu23) [reply egs2/l3das22/enh1/local/data.sh:40]: Added downloading the data from kaggle.

File: egs2/l3das22/enh1/pyscripts (hunk 11 above)
Comment (neillu23) [reply egs2/l3das22/enh1/pyscripts]: Changed.

File: egs2/l3das22/enh1/local/metric.sh (hunk 10 above)
Comment (neillu23) [reply egs2/l3das22/enh1/local/metric.sh:9]: Added.

**Comment by neillu23** [comment]
Thanks for the comments! I have added the downloading data part in the recipe!

File: egs2/l3das22/enh1/run.sh [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+./enh.sh --audio_format wav \
+    --train_set "${train_set}" \
+    --valid_set "${valid_set}" \
+    --test_sets "${test_sets}" \
+    --fs ${sample_rate} \
+    --ngpu 1 \
+    --spk_num 1 \
+    --enh_config conf/tuning/train_enh_beamformer_mvdr.yaml \
+    --use_dereverb_ref false \
+    --use_noise_ref true \
```
Comment (kamo-naoyuki) [inline egs2/l3das22/enh1/run.sh:24]: I have a question.

In my run, noise.scp was not generated. How can I get it?

Comment (Emrys365) [reply egs2/l3das22/enh1/run.sh:24]: I don't think we have access to the noise data. So it should be set to false here.

Comment (kamo-naoyuki) [reply egs2/l3das22/enh1/run.sh:24]: Okay, thanks!

Comment (neillu23) [reply egs2/l3das22/enh1/run.sh]: Yes, I have changed use_noise_ref to false. Thanks for the help!

**Comment by sw005320** [comment]
@neillu23, some comments still exist, including the results.

File: egs2/l3das22/enh1/local/l3das22_multi_channel.sh (hunk 1 above)
Comment (neillu23) [reply egs2/l3das22/enh1/local/l3das22_multi_channel.sh:30]: Thanks! Added in 
https://github.com/neillu23/espnet/blob/l3das22/egs2/l3das22/enh1/local/l3das22_multi_channel.sh

**Comment by neillu23** [comment]
> @neillu23, some comments still exist, including the results.

Thanks for reminding me! I haven't changed the data/temp to `mktemp` in local/l3das22_data_prep.sh because the files under the temp directory are used in the following script (local/l3das22_multi_channel.sh). Do you think we should pass the temp directory variable across the scripts?
I am training a fasnet and I will update the results soon.

File: egs2/l3das22/enh1/local/metric.py [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+import jiwer
+import numpy as np
+import soundfile as sf
+import torch
+import torchaudio
+import transformers
+from pystoi import stoi
+from tqdm import tqdm
+from transformers import Wav2Vec2ForMaskedLM, Wav2Vec2Tokenizer
+
```
Comment (Emrys365) [inline egs2/l3das22/enh1/local/metric.py:22]: Could you apply `isort` to sort the imports?

**Review by Emrys365 — APPROVED** [review]
LGTM in general. I think we can merge it after existing comments are addressed.

**Comment by Emrys365** [comment]
LGTM! I will merge it. Thanks a lot @neillu23
